// Some engineers at Google have come to the conclusion that using "required"
// does more harm than good; they prefer to use only "optional" and "repeated".
// However, this view is not universal. TODO
// Source: https://developers.google.com/protocol-buffers/docs/cpptutorial

syntax = "proto2";

package clockwork;

enum RequestType {
  REQ_UPLOAD_MODEL = 1;
  REQ_INFERENCE = 2;
  REQ_LOAD_REMOTE_MODEL = 3;
  REQ_EVICT = 4;

  RSP_UPLOAD_MODEL = 101;
  RSP_INFERENCE = 102;
  RSP_LOAD_REMOTE_MODEL = 103;
  RSP_EVICT = 104;
}

message RequestHeaderProto {
  optional int32 user_id = 1;
  optional int32 user_request_id = 2;
}

message ResponseHeaderProto {
  optional int32 user_request_id = 1;
  optional int32 status = 2;
  optional string message = 3;
}

message ModelUploadReqProto {
  required RequestHeaderProto header = 1;
  required uint32 batchsize = 3;
  required uint64 so_len = 4;
  required uint64 clockwork_len = 5;
  required uint64 params_len = 6;
}

message ModelUploadRspProto {
  required ResponseHeaderProto header = 1;
  required uint32 model_id = 2;
  required uint64 input_size = 3;
  required uint64 output_size = 4;
}


message ModelInferenceReqProto {
  required RequestHeaderProto header = 1;
  required uint32 model_id = 2;
  required uint32 batch_size = 3;
}

message ModelInferenceRspProto {
  required ResponseHeaderProto header = 1;
  required uint32 model_id = 2;
  required uint32 batch_size = 3;
}

message EvictReqProto {
  required RequestHeaderProto header = 1;
  required uint32 model_id = 2;
}

message EvictRspProto {
  required ResponseHeaderProto header = 1;
}

message LoadModelFromDiskReqProto {
  required RequestHeaderProto header = 1;
  required string remote_path = 2;
}

message LoadModelFromDiskRspProto {
  required ResponseHeaderProto header = 1;
  required uint32 model_id = 2;
  required uint64 input_size = 3;
  required uint64 output_size = 4;
}

enum ActionType {
  ACT_LOAD_MODEL_FROM_DISK = 1;
  ACT_LOAD_WEIGHTS = 2;
  ACT_INFER = 3;
  ACT_EVICT_WEIGHTS = 4;
  ACT_CLEAR_CACHE = 5;
  ACT_GET_WORKER_STATE = 6;

  RES_ERROR = 100;
  RES_LOAD_MODEL_FROM_DISK = 101;
  RES_LOAD_WEIGHTS = 102;
  RES_INFER = 103;
  RES_EVICT_WEIGHTS = 104;
  RES_CLEAR_CACHE = 105;
  RES_GET_WORKER_STATE = 106;
}

message ErrorResultProto {
  required int32 action_id = 1;
  required int32 action_type = 2;
  required int32 status = 3;
  required string message = 4;
}

message TimingProto {
  required int32 begin = 1;
  required uint64 end = 2;
  required uint64 duration = 3;
}

message LoadModelFromDiskActionProto {
  required int32 action_id = 1;
  required int32 model_id = 2;
  required string model_path = 3;
  required uint64 earliest = 4;
  required uint64 latest = 5;
}

message LoadModelFromDiskResultProto {
  required int32 action_id = 1;
  required TimingProto timing = 2;
  required uint64 input_size = 3;
  required uint64 output_size = 4;
  repeated uint32 supported_batch_sizes = 5;
  required uint64 weights_size_in_cache = 6;
}

message LoadWeightsActionProto {
  required int32 action_id = 1;
  required int32 model_id = 2;
  required uint32 gpu_id = 3;
  required uint64 earliest = 4;
  required uint64 latest = 5;
  required uint64 expected_duration = 6;
}

message LoadWeightsResultProto {
  required int32 action_id = 1;
  required TimingProto timing = 2;
}

message EvictWeightsActionProto {
  required int32 action_id = 1;
  required int32 model_id = 2;
  required uint32 gpu_id = 3;
  required uint64 earliest = 4;
  required uint64 latest = 5;
}

message EvictWeightsResultProto {
  required int32 action_id = 1;
  required TimingProto timing = 2;
}

message InferActionProto {
  required int32 action_id = 1;
  required int32 model_id = 2;
  required uint32 gpu_id = 3;
  required uint64 earliest = 4;
  required uint64 latest = 5;
  required uint64 expected_duration = 6;
  required uint32 batch_size = 7;
}

message InferResultProto {
  required int32 action_id = 1;
  required TimingProto copy_input_timing = 2;
  required TimingProto exec_timing = 3;
  required TimingProto copy_output_timing = 4;
}

message ClearCacheActionProto {
  required int32 action_id = 1;
}

message ClearCacheResultProto {
  required int32 action_id = 1;
}

message GetWorkerStateActionProto {
  required int32 action_id = 1;
}

message ModelInfoProto {
  required int32 id = 1;
  required uint32 size = 2; // bytes
}

message WorkerMemoryInfoProto {
  required uint64 weights_cache_total = 1; // bytes
  required uint64 weights_cache_remaining = 2; // bytes
  required uint64 io_pool_total = 3; // bytes
  required uint64 io_pool_remaining = 4; // bytes
  required uint64 workspace_pool_total = 5; // bytes
  required uint64 workspace_pool_remaining = 6; // bytes
  repeated ModelInfoProto models = 7;
}

message GetWorkerStateResultProto {
  required int32 action_id = 1;
  required WorkerMemoryInfoProto worker_memory_info = 2;
}
